patches/components-sync-syncable-write_transaction.cc.patch

```diff
@@ -0,0 +1,12 @@
+diff --git a/components/sync/syncable/write_transaction.cc b/components/sync/syncable/write_transaction.cc
+index 00b248a93830bbafae831d46d1089a0844c57805..2b9a153e9f55b51d22e5568b469add733921b92a 100644
+--- a/components/sync/syncable/write_transaction.cc
++++ b/components/sync/syncable/write_transaction.cc
+@@ -28,6 +28,7 @@ WriteTransaction::WriteTransaction(const base::Location& from_here,
+ }
+ 
+ WriteTransaction::~WriteTransaction() {
++  if (!keep_)
+   delete transaction_;
```

Comment:
this seems like it could cause a memory leak? Can you explain this one?

Reply:
It is used here https://github.com/brave/brave-core/pull/2010/files#diff-db556f7f2868f56a122f18f7e1a1b1b0R71 
(originally from @AlexeyBarabash commit https://github.com/brave/brave-core/commit/978f894e149122b67af5c8ce94ff210d6f424caf#diff-85c67f0ad37d3b33670fcc822c5ab0cf)

The `BaseTransaction` we get from `ApplyChangesFromSyncModel` is a `ReadTransaction` which will not close the wrapped transaction. https://cs.chromium.org/chromium/src/components/sync/engine_impl/sync_manager_impl.cc?l=726&rcl=55129cbc03cd13fd15ffd3838c0499ca2005f019
And we need a WriteTransaction here to get sync nodes synchronized with bookmark nodes.
The wrapped transaction here is already deleted
```
...
10  libc++.dylib                        0x0000000109a8aeb5 operator delete(void*) + 21
11  libchrome_dll.dylib                 0x0000000119f40078 syncer::syncable::WriteTransaction::~WriteTransaction() + 40
12  libchrome_dll.dylib                 0x0000000119f47413 syncer::WriteTransaction::~WriteTransaction() + 67
13  libchrome_dll.dylib                 0x0000000119f47445 syncer::WriteTransaction::~WriteTransaction() + 21
14  libchrome_dll.dylib                 0x000000011bc1ce97 sync_bookmarks::BookmarkChangeProcessor::MoveSyncNode(int, bookmarks::BookmarkNode const*, syncer::BaseTransaction const*) + 663
15  libchrome_dll.dylib                 0x000000011bc21cdc sync_bookmarks::BookmarkChangeProcessor::ApplyChangesFromSyncModel(syncer::BaseTransaction const*, long long, syncer::Immutable<std::__1::vector<syncer::ChangeRecord, std::__1::allocator<syncer::ChangeRecord> >, syncer::DefaultImmutableTraits<std::__1::vector<syncer::ChangeRecord, std::__1::allocator<syncer::ChangeRecord> > > > const&) + 3916
16  libchrome_dll.dylib                 0x0000000119d96f6c syncer::SyncBackendRegistrar::OnChangesApplied(syncer::ModelType, long long, syncer::BaseTransaction const*, syncer::Immutable<std::__1::vector<syncer::ChangeRecord, std::__1::allocator<syncer::ChangeRecord> >, syncer::DefaultImmutableTraits<std::__1::vector<syncer::ChangeRecord, std::__1::allocator<syncer::ChangeRecord> > > > const&) + 92
17  libchrome_dll.dylib                 0x0000000119e3e2bf syncer::SyncManagerImpl::HandleTransactionEndingChangeEvent(syncer::Immutable<syncer::syncable::WriteTransactionInfo, syncer::DefaultImmutableTraits<syncer::syncable::WriteTransactionInfo> > const&, syncer::syncable::BaseTransaction*) + 639
18  libchrome_dll.dylib                 0x0000000119e3e88c non-virtual thunk to syncer::SyncManagerImpl::HandleTransactionEndingChangeEvent(syncer::Immutable<syncer::syncable::WriteTransactionInfo, syncer::DefaultImmutableTraits<syncer::syncable::WriteTransactionInfo> > const&, syncer::syncable::BaseTransaction*) + 60
19  libchrome_dll.dylib                 0x0000000119f3f6d8 syncer::syncable::WriteTransaction::NotifyTransactionChangingAndEnding(syncer::Immutable<std::__1::map<long long, syncer::syncable::EntryKernelMutation, std::__1::less<long long>, std::__1::allocator<std::__1::pair<long long const, syncer::syncable::EntryKernelMutation> > >, syncer::DefaultImmutableTraits<std::__1::map<long long, syncer::syncable::EntryKernelMutation, std::__1::less<long long>, std::__1::allocator<std::__1::pair<long long const, syncer::syncable::EntryKernelMutation> > > > > const&) + 632
20  libchrome_dll.dylib                 0x0000000119f3f3ce syncer::syncable::WriteTransaction::UnlockAndNotify(syncer::Immutable<std::__1::map<long long, syncer::syncable::EntryKernelMutation, std::__1::less<long long>, std::__1::allocator<std::__1::pair<long long const, syncer::syncable::EntryKernelMutation> > >, syncer::DefaultImmutableTraits<std::__1::map<long long, syncer::syncable::EntryKernelMutation, std::__1::less<long long>, std::__1::allocator<std::__1::pair<long long const, syncer::syncable::EntryKernelMutation> > > > > const&) + 94
21  libchrome_dll.dylib                 0x0000000119f3ff9b syncer::syncable::WriteTransaction::~WriteTransaction() + 283
22  libchrome_dll.dylib                 0x0000000119f40045 syncer::syncable::WriteTransaction::~WriteTransaction() + 21
23  libchrome_dll.dylib                 0x0000000119dc704b syncer::DirectoryUpdateHandler::ApplyUpdatesImpl(syncer::StatusController*) + 1435
...

```
If we delete it, that will cause double free. And we can't generate a new WriteTransaction here because it requires a new lock which will cause resource deadlock

Maybe the naming is too confusing, we can use `close_transaction_` as ReadTransaction does 
https://cs.chromium.org/chromium/src/components/sync/syncable/read_transaction.cc?l=26&rcl=3fafe5c5270848088f2677f490157bab3ab7bbbe

Reply:
This is only needed when getting records from old device which doesn't send `position_in_parent` in meta info. With `position_in_parent`, we can have `UniquePosition` calculated before making changes to sync node. For those legacy device, this is the only place we can calculate where to put this bookmark.